### PR TITLE
Switch from ansi-wl-pprint to the prettyprinter package.

### PR DESCRIPTION
**Pull request by brianhuffman**
Note: This PR builds on GaloisInc/what4#77, GaloisInc/saw-core#106, #584, and #585; it will stay in draft status until those PRs are accepted and merged.

EDIT: GaloisInc/saw-core#106, #584, and #585 have all been merged, so we're just waiting on GaloisInc/what4#77.

### CODE REVIEW

File: crucible-llvm/src/Lang/Crucible/LLVM/MemType.hs [hunk 1]
```diff
@@ -79,8 +79,9 @@ binarySearch f = go
                                GT -> go l i
           where i = l + (h - l) `div` 2
 
-ppIdent :: L.Ident -> Doc
-ppIdent = text . show . L.ppIdent
+ppIdent :: L.Ident -> Doc ann
+ppIdent = pretty . show . L.ppIdent
+-- TODO: update if llvm-pretty switches to prettyprinter
```
Comment (kquick) [inline crucible-llvm/src/Lang/Crucible/LLVM/MemType.hs:84]: Yay! :-)

File: crux-mir/src/Mir/PP.hs [hunk 2]
```diff
@@ -235,86 +239,86 @@ instance Pretty Constant where
...
 instance Pretty Substs where
   pretty (Substs b) = langle <> hcat (punctuate comma (map pretty b)) <> rangle
-  
+
 instance Pretty ConstVal where
     pretty (ConstFloat i)   = pretty i
     pretty (ConstInt i)     = pretty i
-    pretty (ConstStr i)     = char '\"' <> text (show i) <> char '\"'
-    pretty (ConstByteStr i) = text (show i)
+    pretty (ConstStr i)     = pretty '\"' <> pretty (show i) <> pretty '\"'
```
Comment (kquick) [inline crux-mir/src/Mir/PP.hs:302]: Maybe `pretty (ConstStr i) = dquotes (pretty i)` and `pretty (ConstByteStr i) = pretty i`

Simplifies the quoting, and gets rid of the show on the internal (if possible).

File: crux-mir/src/Mir/Trans.hs [hunk 3]
```diff
@@ -323,7 +324,7 @@ extendToMax n e1 m e2 (Just arith) k =
                           Just LeqProof ->
                               k n e1 (S.app $ extend n m e2)
                           Nothing -> error "impossible case"
-extendToMax n e1 m e2 Nothing k = 
+extendToMax n e1 m e2 Nothing k =
```
Comment (kquick) [inline crux-mir/src/Mir/Trans.hs:327]: Just being picky, but I usually try to make unrelated changes line this in different PR's.

**Review by kquick — APPROVED** [review]
In about a half-dozen places you fixed the annotation to `()`; with the parameterized `ann` the rest of the time that seems like a shortest-path conversion but doesn't enhance the context for later handling like colorization.  If the "shortest-path" assumption is correct, it might be nice to note that in the description of the PR for future reference (as opposed to having a reason for it explicitly being `()`).

There were lots of locations that were effectively `pretty . show`, and you marked a couple but not many of those.  It would be great to shorten those to `pretty` where possible and mark them where not possible with some kind of consistent comment mark for future updating.

Also not a lot of consistency between leveraging `IsString` and using `pretty`; if this was mostly a mechanical `sed`-style change, it might be worth going through and standardizing on the former where possible.

Mostly an eyeball-glazing mechanical set of changes, thanks for doing all of this! I had a few comments above and throughout but nothing significant enough to prevent merging, so feel free to ignore or change as you'd prefer but I think this is a good step forward to get merged.

File: crucible-llvm/src/Lang/Crucible/LLVM/Errors/MemoryError.hs [hunk 4]
```diff
@@ -78,95 +77,98 @@ data MemoryErrorReason sym w =
...
+       , indent 2 (hsep ([ "Source:     " ] ++ ppGSym gsym_src ++ [ ppPtr src ]))
+       , "In memory state:"
        , indent 2 (ppMem mem)
        ]
 
 ppMemoryOp (MemLoadHandleOp sig gsym ptr mem) =
-  vsep [ "Attempting to load callable function with type:" <+> text (show (L.ppType sig))
-       , indent 2 (hsep ([ text "Via pointer:" ] ++ ppGSym gsym ++ [ ppPtr ptr ]))
-       , text "In memory state:"
+  vsep [ "Attempting to load callable function with type:" <+> pretty (show (L.ppType sig))
```
[the remark this answers is not in the document]
Comment (brianhuffman) [reply crucible-llvm/src/Lang/Crucible/LLVM/Errors/MemoryError.hs:127]: The `llvm-pretty` library actually uses `Text.PrettyPrint.HughesPJ` from the `pretty` package. I suppose we should make a PR for `llvm-pretty` to merge before this one, and then we can shortcut some of these `pretty`/`show` combinations. I did mark some occurrences of this pattern with comments, but there were getting to be a few too many of them.

File: crux-mir/src/Mir/PP.hs (hunk 2 above)
Comment (brianhuffman) [reply crux-mir/src/Mir/PP.hs:302]: I would have combined the `pretty` and `show` on those if I had known I could. The `i` on both of those constructors has type `ByteString`, and I actually don't know exactly how `Show` instance is defined for that, and I didn't check to see if there was already a `Pretty` instance. If there is, and they match, then I could go back and combine those.

File: crux-mir/src/Mir/Trans.hs (hunk 3 above)
Comment (brianhuffman) [reply crux-mir/src/Mir/Trans.hs:327]: There were trailing spaces on some of the lines I was editing, so I just ran `delete-trailing-spaces` on the whole file.

**Comment by brianhuffman** [comment]
About using `()` as an annotation in some places: I had some doubts about that. None of the code anywhere used any color, ANSI features or other annotations. So I thought the best choice was to make all the pretty printing code polymorphic in the annotation type as much as possible, to make it easy to combine those pretty printers with ones from other packages.

One stumbling block was that there are a few datatypes that stored `Doc` values inside. What annotation type should I use for that? One possibility is to add an `ann` parameter to the datatype, which I did in one or two places in GaloisInc/what4#77 (mostly this was for pretty-printing-related datatypes, where it seemed to make sense). Another possibility would be to use rank-2 polymorphism, with a constructor field of type `forall ann. Doc ann`, but this seemed like overkill. The last possibility is to use a dummy type that basically signals that you're not using any annotations; I considered both `()` and `Void`. Actually `Doc Void` is probably the right choice semantically, as it basically doesn't allow you to use any annotations, but `Doc ()` seemed to convey the intention better. Other places that use `Doc ()` are mostly just the minimum necessary to get things to typecheck. I may have used `Doc ()` in a couple of other places that use `Doc` only on input parameters, just to avoid ambiguous types (although, now that I think about it, I don't think GHC even has a problem with this kind of ambiguity if there's no class constraint on the variable, so maybe I can get rid of a few of these).

I'm not sure what we should consider the best practice for handling the `Doc`-inside-a-datatype thing; I couldn't find any relevant advice in the package documentation.

File: crux-mir/src/Mir/PP.hs (hunk 2 above)
Comment (brianhuffman) [reply crux-mir/src/Mir/PP.hs:302]: I just looked through some haddocks. The `prettyprinter` library doesn't provide a `Pretty` instance for `ByteString`. The `Show` instance for `ByteString` is completely undocumented, but from peeking at the source, `show` is the same as `show . Data.ByteString.Char8.unpack`, so it's like printing a `String` where every `Char` fits in 8 bits.

So this code seems a bit weird, then. As far as I can tell, a `ConstByteStr` will print out as a quoted string (just like calling `show` on a `String`) and a `ConstStr` will print out as a string with double-double quotes around it, like `""foo""`. I don't know if this is what was intended here or not.

**Comment by brianhuffman** [comment]
My latest thinking is that we should try to completely avoid `pretty (show x)`, other than to define `Pretty` instances or custom type-specific pretty printing functions. Every occurrence of `pretty (show x)` is technical debt, in that we would prefer to get rid of it as soon as possible once other libraries get the required pretty printing support. So we should try to factor out all such occurrences into a small set of top-level functions, to make that later process easier.

File: crux-mir/src/Mir/Trans.hs (hunk 3 above)
Comment (kquick) [reply crux-mir/src/Mir/Trans.hs:327]: I love ws-butler-mode (in emacs): it automatically removes extraneous whitespace around the areas you are editing, but leaves the rest of the file alone.  This obviously doesn't fully convert the entire file, but it does avoid introducing these un-related changes into your current PR whilst ensuring the portions you've worked on are in good shape.

**Comment by kquick** [comment]
I agree that it's unwanted tech-debt, but it's going to be hard to completely eliminate these due to using external libraries (even our own) that don't have prettyprinter `Doc` formatting.

One suggestion is to use the prettyprinter `viaShow` function when we can't eliminate this tech-debt: this function makes it clear that the pretty-printing was generated from a `Show` instance and is a nice footprint for future search/replace tech-debt reduction.

**Comment by brianhuffman** [comment]
I just came across `viaShow` and I agree that it would be a good solution to make it easier to find places to fix later. So I think there's really no excuse for using `pretty . show` at all.

**Comment by robdockins** [comment]
> About using `()` as an annotation in some places: I had some doubts about that. None of the code anywhere used any color, ANSI features or other annotations. So I thought the best choice was to make all the pretty printing code polymorphic in the annotation type as much as possible, to make it easy to combine those pretty printers with ones from other packages.
> 
> One stumbling block was that there are a few datatypes that stored `Doc` values inside. What annotation type should I use for that? One possibility is to add an `ann` parameter to the datatype, which I did in one or two places in [GaloisInc/what4#77](https://github.com/GaloisInc/what4/pull/77) (mostly this was for pretty-printing-related datatypes, where it seemed to make sense). Another possibility would be to use rank-2 polymorphism, with a constructor field of type `forall ann. Doc ann`, but this seemed like overkill. The last possibility is to use a dummy type that basically signals that you're not using any annotations; I considered both `()` and `Void`. Actually `Doc Void` is probably the right choice semantically, as it basically doesn't allow you to use any annotations, but `Doc ()` seemed to convey the intention better. Other places that use `Doc ()` are mostly just the minimum necessary to get things to typecheck. I may have used `Doc ()` in a couple of other places that use `Doc` only on input parameters, just to avoid ambiguous types (although, now that I think about it, I don't think GHC even has a problem with this kind of ambiguity if there's no class constraint on the variable, so maybe I can get rid of a few of these).
> 
> I'm not sure what we should consider the best practice for handling the `Doc`-inside-a-datatype thing; I couldn't find any relevant advice in the package documentation.

I think we should use `Void` as the annotation type anyplace where we are currently not using annotations.  It accurately reflects the intention not to use them, and `Void` is easily remapped into whatever annotation you like in the places where that becomes necessary.

**Comment by kquick** [comment]
Re-reviewed with all the updated changes: this looks good to me!
